src/platform/ASR/ASRFactoryDataParser.c

```diff
@@ -0,0 +1,308 @@
+    },
+    [ASR_HARDWARE_VERSION_STR_PARTITION] =
+    {
+        .partition_name = "hw-ver-str",
+    },
+    [ASR_PRODUCT_URL_PARTITION] =
+    {
+        .partition_name = "product-url",
+    },
+    [ASR_PRODUCT_LABEL_PARTITION] =
+    {
+        .partition_name = "product-label",
+    },
+    [ASR_PART_NUMBER_PARTITION] =
+    {
+        .partition_name = "part-number",
+    },
+#endif
+    [ASR_MATTER_PARTITION_MAX] =
+    {
+        .partition_name = NULL, //for end don't change,
+    }
+};
+
+static asr_tlv_context matter_tlv_ctx;
+
+uint32_t asr_matter_find_by_name(const char * name, uint8_t * buf, uint32_t buf_len, uint32_t * out_len)
+{
+    tlv_header_t tlv;
+
+    if (asr_tlv_find_by_name(&matter_tlv_ctx, &tlv, MIXCLASS_MATTERCONFIG_TAG,
+                             (char *) name)) // parse the data according to the tlv packaging tool
+    {
+        uint32_t value_len = tlv_htons(tlv->data_len);
+
+        value_len -= MAX_NAME_LEN;
+
+        if (value_len > buf_len)
+        {
+            // buf_len too small
+            return -2;
+        }
+
+        *out_len = value_len;
+
+        memcpy(buf, tlv->data + MAX_NAME_LEN, value_len);
+
+        return 0;
+    }
+    else
+    {
+        return -1;
+    }
+}
+
+static uint32_t asr_factory_toltal_len()
+{
+    tlv_area_header_t tlv_headr_p = (tlv_area_header_t) MATTER_FLASH_START_ADDR;
+    return tlv_headr_p->data_len + 3 * sizeof(uint32_t); // magic_num,crc32_value,data_len
+}
+
+#if defined(CFG_PLF_RV32) || defined(CFG_PLF_DUET)
+extern int32_t asr_factory_decrypt_dac_prvkey(uint8_t * pDacCipher, uint8_t ucDacKeyLen, uint8_t * pRdBuf, uint32_t * pOutLen);
+#endif
+static uint32_t matter_config_value = 0;
+static int32_t asr_factory_dac_prvkey_get(uint8_t * pRdBuf, uint32_t * pOutLen)
+{
+#if defined(CFG_PLF_RV32) || defined(CFG_PLF_DUET)
+    uint32_t off_set = asr_factory_toltal_len();
+    uint8_t ucCipher[DAC_PRIKEY_LEN];
+#endif
+    int32_t ret = 0;
+
+    if (NULL == pRdBuf || NULL == pOutLen)
+    {
+        return -1;
+    }
+    if (matter_config_value & MATTER_NO_KEY)
+    {
+#if defined(CFG_PLF_RV32) || defined(CFG_PLF_DUET)
+        // get DAC cipher
+        ret = asr_flash_read(ASR_CONFIG_BASE, (uint32_t *) &off_set, (void *) ucCipher, DAC_PRIKEY_LEN);
+        if (ret == 0)
+        {
+            ret = asr_factory_decrypt_dac_prvkey(ucCipher, 32, pRdBuf, pOutLen);
+        }
+#endif
+    }
+    else
+    {
+        ret = asr_matter_find_by_name(asr_matter_partitions_table[ASR_DAC_KEY_PARTITION].partition_name, pRdBuf, DAC_PRIKEY_LEN,
+                                      pOutLen);
+    }
+    return ret;
+}
+
+int32_t asr_factory_config_read(asr_matter_partition_t matter_partition, uint8_t * buf, uint32_t buf_len, uint32_t * out_len)
+{
+    int32_t ret = 0;
+    if (matter_partition < ASR_MATTER_PARTITION_MAX)
+    {
+        if (matter_partition == ASR_DAC_KEY_PARTITION)
+        {
+            ret = asr_factory_dac_prvkey_get(buf, out_len);
```

Comment:
Reading through this, we return ret if not 0, otherwise we continue and ... return ret. How about:

```cpp
if (matter_partition >= ASR_MATTER_PARTITION_MAX)
{
    return -3; // why the magic number? can we have a constant?
}

if (matter_partition == ASR_DAC_KEY_PARTITION)
{
     return asr_factory_dac_prvkey_get(buf, out_len);
 } 

return asr_matter_find_by_name(asr_matter_partitions_table[matter_partition].partition_name, buf, buf_len, out_len);
```

Seems more compact and more linear to read